File: src/contract/types/passives.ts
// ─────────────────────────────────────────────────────────────────────────
// 특성(trait)/패시브(passive) 룰 스키마 (디자이너 작성). when/if/then 합성 언어.
// 엔진(core/combat/passives)이 이걸 "해석"만 한다. (8.6 / 8.8 데이터·엔진 경계)
// 트리거는 현재 전투 스코프 전체. 모험 스코프(노드/골드/파티HP)는 후속 슬라이스.
// ─────────────────────────────────────────────────────────────────────────
import type { NodeType, Side } from "./content.ts";

export type Comparator = "lt" | "lte" | "eq" | "gte" | "gt";

/** 트리거 대상 필터. self=규칙 소유자 / ally·enemy=소유자 기준 진영 / any=전체 */
export type TriggerWho = "self" | "ally" | "enemy" | "any";

/** 발동 시점 (when). 판별 유니온 `on`. */
export type Trigger =
  // 턴/라운드
  | { on: "battleStart" }
  | { on: "roundStart" }
  | { on: "roundEnd" }
  | { on: "turnStart"; who?: TriggerWho }
  | { on: "turnEnd"; who?: TriggerWho }
  | { on: "everyNTurns"; n: number } // 소유자의 자기 턴 카운터 기준
  | { on: "interruptStart" }
  // 주사위 (effect는 modSpeedRoll/rerollSpeed만 유효)
  | { on: "speedRoll" }
  // 행동/스킬/이동
  | { on: "beforeAction"; who?: TriggerWho }
  | { on: "skillUsed"; who?: TriggerWho; skillId?: string }
  | { on: "onMove"; who?: TriggerWho }
  | { on: "enterCell"; row?: number; col?: number }
  // 명중/피해/회복/쉴드
  | { on: "onHit"; as?: "attacker" | "target"; crit?: boolean }
  | { on: "onMiss"; as?: "attacker" | "target" }
  | { on: "dealtDamage" } // 소유자가 가해자
  | { on: "damaged" } // 소유자가 피격자
  | { on: "onHeal"; as?: "healer" | "target" }
  | { on: "onShieldGain" } // 소유자가 쉴드 획득
  // 상태/사망
  | { on: "statusApplied"; statusId?: string; as?: "applier" | "target" }
  | { on: "statusTick"; statusId?: string } // on-bleed 등 (지속 피해/회복 발동 시)
  | { on: "kill" } // 소유자가 처치
  | { on: "death"; who?: TriggerWho }
  | { on: "battleEnd"; result?: "win" | "lose" }
  // ── 모험(run) 스코프 — core/run/passives.ts가 해석 ──
  | { on: "nodeEnter"; nodeType?: NodeType }
  | { on: "nodeClear"; nodeType?: NodeType }
  | { on: "actStart" }
  | { on: "goldGain" }
  | { on: "partyHpChange"; dir?: "heal" | "hurt" };

/** 조건 평가 대상. self=소유자 / subject=이벤트 1차 대상(피격자 등) / target=소유자 현재 행동 대상 */
export type CondWho = "self" | "subject" | "target";

/** 조건 (if). AND 결합, 비면 항상 참. */
export type Condition =
  | { c: "hpPct"; who: CondWho; cmp: Comparator; v: number }
  | { c: "round"; cmp: Comparator; v: number }
  | { c: "selfTurnCount"; cmp: Comparator; v: number }
  | { c: "everyN"; n: number; of: "round" | "selfTurn" }
  | { c: "firstTurn" }
  | { c: "hasStatus"; who: CondWho; statusId: string; minStacks?: number }
  | { c: "missingStatus"; who: CondWho; statusId: string }
  | { c: "atColumn"; who: CondWho; cmp: Comparator; v: number }
  | { c: "atRow"; who: CondWho; cmp: Comparator; v: number }
  | { c: "atCell"; who: CondWho; row: number; col: number }
  | { c: "isFrontline"; who: CondWho } // 점유 최전열(reach 기준)에 있는가
  | { c: "sideCount"; side: Side; cmp: Comparator; v: number } // 살아있는 수
  | { c: "outnumbered" } // 아군 살아있는 수 < 적
  | { c: "subjectCharId"; charId: string }
  | { c: "subjectSide"; side: Side }
  | { c: "wasCrit" }
  | { c: "damageAtLeast"; v: number }
  | { c: "skillIs"; skillId: string }
  | { c: "chance"; pct: number } // state.rng만 사용
  // ── 모험(run) 스코프 조건 ──
  | { c: "nodeTypeIs"; nodeType: NodeType }
  | { c: "goldAtLeast"; v: number };

/** 효과 대상. other*=소유자/대상 제외 광역. */
export type EffTarget = "self" | "subject" | "target" | "allAllies" | "allEnemies" | "otherAllies" | "otherEnemies" | "randomEnemy" | "randomAlly";
/** statMod 가능한 스탯. */
export type StatKey = "accuracy" | "evasion" | "critChance" | "critMultiplier" | "speedMin" | "speedMax";

/** 효과 (then). 순서대로 실행. */
export type Effect =
  // 기존 전투 프리미티브 재사용
  | { do: "damage"; amount: number; target: EffTarget }
  | { do: "heal"; amount: number; target: EffTarget }
  | { do: "shield"; amount: number; target: EffTarget }
  | { do: "applyStatus"; statusId: string; stacks: number; duration: number; target: EffTarget }
  | { do: "cleanse"; target: EffTarget }
  | { do: "move"; deltaCol: number; target: EffTarget }
  | { do: "grantInterrupt"; count: number; target: EffTarget }
  // 신규 해석
  | { do: "modSpeedRoll"; delta: number } // speedRoll 트리거 전용(소유자 주사위에 가산, rng 호출 없음)
  | { do: "rerollSpeed" } // speedRoll 트리거 전용(소유자 주사위 재굴림, state.rng)
  | { do: "statMod"; stat: StatKey; delta: number; target: EffTarget } // 전투 동안 누적 스탯 보정
  | { do: "modCooldown"; skillId?: string; delta: number; target: EffTarget } // 쿨다운 가감(음수=감소)
  // ── 피해 비례/상태 제거 (전투 스코프 — dealtDamage/damaged 등 ctx.damage 필요) ──
  | { do: "healByDamage"; pct: number; target: EffTarget } // 흡혈: 가한 피해의 pct% 회복
  | { do: "reflectByDamage"; pct: number; target: EffTarget } // 비율 반사: 받은 피해의 pct%를 대상에 피해
  | { do: "removeStatus"; statusId: string; target: EffTarget } // 특정 상태이상 1종 제거
  | { do: "castSkill"; skillId: string } // 액티브 스킬을 정상 시전(명중·사정권·면적은 스킬 정의대로, 타겟 자동). 대상은 passives 없는 leaf 스킬만(재귀 방지)
  // ── 연출(스토리텔링) — phase/상태 불변, 뷰 이벤트만 (NODE-DESIGN Phase C) ──
  | { do: "showDialog"; speaker?: string; text: string } // 대사/컷신 1줄 → dialog 이벤트 push(전투 위 오버레이로 재생)
  // ── 모험(run) 스코프 효과 ──
  | { do: "goldDelta"; amount: number } // 골드 가감
  | { do: "healParty"; pct: number } // 파티 비율 회복
  | { do: "grantRunStatus"; statusId: string; stacks: number; duration: number; target: EffTarget }; // 다음 전투 시작 시 부여(계승)

/** 디자이너 룰: when 시점에, if 모두 참이면, then 순서대로 실행. */
export interface PassiveRule {
  id?: string;
  when: Trigger;
  if?: Condition[];
  then: Effect[];
  maxPerTurn?: number; // 소유자 턴당 최대 발동(과발동 방지)
  maxPerBattle?: number; // 전투당 최대 발동
}

/** 특성 = 캐릭터를 정의하는 패시브 룰 묶음(data/traits.ts). 캐릭터가 traitIds로 참조. */
export interface TraitDef {
  id: string;
  name: string;
  icon?: string;
  desc?: string;
  rules: PassiveRule[];
}

File: rust/spr-types/src/passives.rs
//! 특성/패시브 DSL 스키마 (TS `types/passives.ts`). when/if/then 합성 언어. 엔진이 해석만.
//! 판별자: Trigger=`on`, Condition=`c`, Effect=`do` (DATA-SERIALIZATION-CONTRACT). who/as 등은 String.
use serde::{Deserialize, Serialize};

/// 발동 시점(when). 판별 `on`. who/as/result/dir 등은 Option<String>(TS 문자열 비교 그대로).
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "on")]
pub enum Trigger {
    #[serde(rename = "battleStart")]
    BattleStart,
    #[serde(rename = "roundStart")]
    RoundStart,
    #[serde(rename = "roundEnd")]
    RoundEnd,
    #[serde(rename = "turnStart")]
    TurnStart { who: Option<String> },
    #[serde(rename = "turnEnd")]
    TurnEnd { who: Option<String> },
    #[serde(rename = "everyNTurns")]
    EveryNTurns { n: i64 },
    #[serde(rename = "interruptStart")]
    InterruptStart,
    #[serde(rename = "speedRoll")]
    SpeedRoll,
    #[serde(rename = "beforeAction")]
    BeforeAction { who: Option<String> },
    #[serde(rename = "skillUsed")]
    SkillUsed {
        who: Option<String>,
        #[serde(rename = "skillId")]
        skill_id: Option<String>,
    },
    #[serde(rename = "onMove")]
    OnMove { who: Option<String> },
    #[serde(rename = "enterCell")]
    EnterCell { row: Option<i64>, col: Option<i64> },
    #[serde(rename = "onHit")]
    OnHit {
        #[serde(rename = "as")]
        as_role: Option<String>,
        crit: Option<bool>,
    },
    #[serde(rename = "onMiss")]
    OnMiss {
        #[serde(rename = "as")]
        as_role: Option<String>,
    },
    #[serde(rename = "dealtDamage")]
    DealtDamage,
    #[serde(rename = "damaged")]
    Damaged,
    #[serde(rename = "onHeal")]
    OnHeal {
        #[serde(rename = "as")]
        as_role: Option<String>,
    },
    #[serde(rename = "onShieldGain")]
    OnShieldGain,
    #[serde(rename = "statusApplied")]
    StatusApplied {
        #[serde(rename = "statusId")]
        status_id: Option<String>,
        #[serde(rename = "as")]
        as_role: Option<String>,
    },
    #[serde(rename = "statusTick")]
    StatusTick {
        #[serde(rename = "statusId")]
        status_id: Option<String>,
    },
    #[serde(rename = "kill")]
    Kill,
    #[serde(rename = "death")]
    Death { who: Option<String> },
    #[serde(rename = "battleEnd")]
    BattleEnd { result: Option<String> },
    // 모험(run) 스코프 — 전투 디스패처는 무시(run/passives가 해석).
    #[serde(rename = "nodeEnter")]
    NodeEnter {
        #[serde(rename = "nodeType")]
        node_type: Option<String>,
    },
    #[serde(rename = "nodeClear")]
    NodeClear {
        #[serde(rename = "nodeType")]
        node_type: Option<String>,
    },
    #[serde(rename = "actStart")]
    ActStart,
    #[serde(rename = "goldGain")]
    GoldGain,
    #[serde(rename = "partyHpChange")]
    PartyHpChange { dir: Option<String> },
}

impl Trigger {
    /// 판별 문자열(`on` 값) — 디스패처가 tctx.on과 매칭.
    pub fn on(&self) -> &'static str {
        match self {
            Trigger::BattleStart => "battleStart",
            Trigger::RoundStart => "roundStart",
            Trigger::RoundEnd => "roundEnd",
            Trigger::TurnStart { .. } => "turnStart",
            Trigger::TurnEnd { .. } => "turnEnd",
            Trigger::EveryNTurns { .. } => "everyNTurns",
            Trigger::InterruptStart => "interruptStart",
            Trigger::SpeedRoll => "speedRoll",
            Trigger::BeforeAction { .. } => "beforeAction",
            Trigger::SkillUsed { .. } => "skillUsed",
            Trigger::OnMove { .. } => "onMove",
            Trigger::EnterCell { .. } => "enterCell",
            Trigger::OnHit { .. } => "onHit",
            Trigger::OnMiss { .. } => "onMiss",
            Trigger::DealtDamage => "dealtDamage",
            Trigger::Damaged => "damaged",
            Trigger::OnHeal { .. } => "onHeal",
            Trigger::OnShieldGain => "onShieldGain",
            Trigger::StatusApplied { .. } => "statusApplied",
            Trigger::StatusTick { .. } => "statusTick",
            Trigger::Kill => "kill",
            Trigger::Death { .. } => "death",
            Trigger::BattleEnd { .. } => "battleEnd",
            Trigger::NodeEnter { .. } => "nodeEnter",
            Trigger::NodeClear { .. } => "nodeClear",
            Trigger::ActStart => "actStart",
            Trigger::GoldGain => "goldGain",
            Trigger::PartyHpChange { .. } => "partyHpChange",
        }
    }
}

/// 조건(if). 판별 `c`. cmp는 Comparator 문자열.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "c")]
pub enum Condition {
    #[serde(rename = "hpPct")]
    HpPct { who: String, cmp: String, v: i64 },
    #[serde(rename = "round")]
    Round { cmp: String, v: i64 },
    #[serde(rename = "selfTurnCount")]
    SelfTurnCount { cmp: String, v: i64 },
    #[serde(rename = "everyN")]
    EveryN { n: i64, of: String },
    #[serde(rename = "firstTurn")]
    FirstTurn,
    #[serde(rename = "hasStatus")]
    HasStatus {
        who: String,
        #[serde(rename = "statusId")]
        status_id: String,
        #[serde(rename = "minStacks")]
        min_stacks: Option<i64>,
    },
    #[serde(rename = "missingStatus")]
    MissingStatus {
        who: String,
        #[serde(rename = "statusId")]
        status_id: String,
    },
    #[serde(rename = "atColumn")]
    AtColumn { who: String, cmp: String, v: i64 },
    #[serde(rename = "atRow")]
    AtRow { who: String, cmp: String, v: i64 },
    #[serde(rename = "atCell")]
    AtCell { who: String, row: i64, col: i64 },
    #[serde(rename = "isFrontline")]
    IsFrontline { who: String },
    #[serde(rename = "sideCount")]
    SideCount { side: String, cmp: String, v: i64 },
    #[serde(rename = "outnumbered")]
    Outnumbered,
    #[serde(rename = "subjectCharId")]
    SubjectCharId {
        #[serde(rename = "charId")]
        char_id: String,
    },
    #[serde(rename = "subjectSide")]
    SubjectSide { side: String },
    #[serde(rename = "wasCrit")]
    WasCrit,
    #[serde(rename = "damageAtLeast")]
    DamageAtLeast { v: i64 },
    #[serde(rename = "skillIs")]
    SkillIs {
        #[serde(rename = "skillId")]
        skill_id: String,
    },
    #[serde(rename = "chance")]
    Chance { pct: i64 },
    #[serde(rename = "nodeTypeIs")]
    NodeTypeIs {
        #[serde(rename = "nodeType")]
        node_type: String,
    },
    #[serde(rename = "goldAtLeast")]
    GoldAtLeast { v: i64 },
}

/// 효과(then). 판별 `do`. target은 EffTarget 문자열.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "do")]
pub enum Effect {
    #[serde(rename = "damage")]
    Damage { amount: i64, target: String },
    #[serde(rename = "heal")]
    Heal { amount: i64, target: String },
    #[serde(rename = "shield")]
    Shield { amount: i64, target: String },
    #[serde(rename = "applyStatus")]
    ApplyStatus {
        #[serde(rename = "statusId")]
        status_id: String,
        stacks: i64,
        duration: i64,
        target: String,
    },
    #[serde(rename = "cleanse")]
    Cleanse { target: String },
    #[serde(rename = "move")]
    Move {
        #[serde(rename = "deltaCol")]
        delta_col: i64,
        target: String,
    },
    #[serde(rename = "grantInterrupt")]
    GrantInterrupt { count: i64, target: String },
    #[serde(rename = "modSpeedRoll")]
    ModSpeedRoll { delta: i64 },
    #[serde(rename = "rerollSpeed")]
    RerollSpeed,
    #[serde(rename = "statMod")]
    StatMod { stat: String, delta: i64, target: String },
    #[serde(rename = "modCooldown")]
    ModCooldown {
        #[serde(rename = "skillId")]
        skill_id: Option<String>,
        delta: i64,
        target: String,
    },
    #[serde(rename = "healByDamage")]
    HealByDamage { pct: i64, target: String },
    #[serde(rename = "reflectByDamage")]
    ReflectByDamage { pct: i64, target: String },
    #[serde(rename = "removeStatus")]
    RemoveStatus {
        #[serde(rename = "statusId")]
        status_id: String,
        target: String,
    },
    #[serde(rename = "castSkill")]
    CastSkill {
        #[serde(rename = "skillId")]
        skill_id: String,
    },
    #[serde(rename = "showDialog")]
    ShowDialog { speaker: Option<String>, text: String },
    #[serde(rename = "goldDelta")]
    GoldDelta { amount: i64 },
    #[serde(rename = "healParty")]
    HealParty { pct: i64 },
    #[serde(rename = "grantRunStatus")]
    GrantRunStatus {
        #[serde(rename = "statusId")]
        status_id: String,
        stacks: i64,
        duration: i64,
        target: String,
    },
}

/// 디자이너 룰: when 시점에 if 모두 참이면 then 순서대로. TS PassiveRule.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PassiveRule {
    #[serde(default)]
    pub id: Option<String>,
    pub when: Trigger,
    #[serde(rename = "if", default)]
    pub if_conds: Option<Vec<Condition>>,
    pub then: Vec<Effect>,
    #[serde(rename = "maxPerTurn", default)]
    pub max_per_turn: Option<i64>,
    #[serde(rename = "maxPerBattle", default)]
    pub max_per_battle: Option<i64>,
}

/// 특성 = 캐릭터 정의 패시브 묶음(data/traits). TS TraitDef.
#[derive(Debug, Clone, Deserialize)]
pub struct TraitDef {
    pub id: String,
    pub name: String,
    pub rules: Vec<PassiveRule>,
}
